packages/compose/src/hooks/use-shallow-compare-effect/index.js

```diff
@@ -0,0 +1,26 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useRef } from '@wordpress/element';
+import isShallowEqual from '@wordpress/is-shallow-equal';
+
+/**
+ * Like `useEffect` but call the effect when the dependencies are not shallowly equal.
+ * Useful when the size of the dependency array might change during re-renders.
+ * This hook is only used for backward-compatibility reason. Consider using `useEffect` wherever possible.
+ *
+ * @param {Function} effect The effect callback passed to `useEffect`.
+ * @param {Array}    deps   The dependency array that is compared against shallowly.
+ */
+function useShallowCompareEffect( effect, deps ) {
+	const ref = useRef();
+
+	if ( ! isShallowEqual( ref.current, deps ) ) {
+		ref.current = deps;
+	}
+
+	// eslint-disable-next-line react-hooks/exhaustive-deps
+	useEffect( effect, [ ref.current ] );
```

Comment:
How does this work? When setting `ref.current`, there's no guarantee that the component will re-render, and therefore no guarantee that `useEffect` and the callback will be called.

Reply:
It's an escape hatch, so not intuitive at all 😞.

How it works:
1. When we receive a new `deps` in the render phase, we compare it with the last value we've seen (`ref.current`).
2. If they are shallowly equal, we don't change `ref.current`, so that the `useEffect` below won't run.
3. If they are not shallow equal, we set `ref.current` to the new `deps`. Here we assume that if two `deps` are not shallowly equal, then they are also not referential equal.
4. If `ref.current` changes, the effect will run.

So basically, it's not that `ref.current` triggers the re-render, but the re-render triggers the setting of the new `ref.current`, and therefore triggers the effect to run.

It's essentially how [`use-deep-compare-effect`](https://github.com/kentcdodds/use-deep-compare-effect/blob/master/src/index.ts) is implemented, but with shallow equality rather than deep equality.

FWIW though, it seems like setting `ref.current` in the render phase is now considered a anti-pattern. I don't think we have to change it now though since this hook is mostly for backward-compatibility anyway 🤔 .